reports/spec4-alloy.tex

```diff
@@ -1,7 +1,144 @@
+      allows us to run the latest off-the-shelf solvers such as Kissat, the
+      winner of the SAT Competition 2024~\cite{SAT-Competition-2024-solvers}.
+
+\end{itemize}
+
+Having \texttt{Spec 4} in~\tlap{}, it was relatively easy for us to write down
+the Alloy specification, as \texttt{Spec 4} is already quite abstract. Since
+Alloy's core abstractions are objects and relations between them, the
+specification looks quite differently. For example, here is how we declare
+signatures for the blocks and checkpoints:
+
+\begin{lstlisting}[language=alloy,columns=fullflexible]
+  sig Payload {}
+  sig Signature {}
+  fact atLeastFourSignatures { #Signature >= 4 }
+
+  sig Block {
+    slot: Int,
+    body: Payload,
+    parent: Block
+  }
+  sig Checkpoint {
+    block: Block,
+    slot: Int
+  }
+
+  one sig GenesisPayload extends Payload {}
+  one sig GenesisBlock extends Block {}
+\end{lstlisting}
+
+In contrast to~\texttt{Spec 4}, our Alloy specification does not describe a
+state machine, but it specifies an arbitrary single state of the protocol. As
+in~\texttt{Spec 4}, we employ the same trick to restrict the justified
+checkpoints:
+
+\begin{lstlisting}[language=alloy,columns=fullflexible]
+  one sig JustifiedCheckpoints {
+    justified: set Checkpoint
+  }
+  fact justifiedCheckpointsAreJustified {
+    all c: JustifiedCheckpoints.justified |
+      c.slot = 0 or 3.mul[#justifyingVotes[c].validator] >= 2.mul[#Signature]
+  }
+\end{lstlisting}
+
+Similar to our experiments with Apalache, we produced examples of configurations
+that satisfy simple properties. For example:
+
+\begin{lstlisting}[language=alloy,columns=fullflexible]
+  run someFinalizedCheckpoint { some c: Checkpoint |
+    isFinalized[c] and c.slot != 0
+  }
+  for 10 but 6 Block, 6 Checkpoint, 12 Vote, 5 int
+\end{lstlisting}
+
+For small search scopes, Alloy was quickly finding examples, often in a matter
+of seconds.
+
+Ultimately, we are interested in showing that~$\textit{AccountableSafety}$
+holds true. To this end, we want Alloy to show that the
+formula~$\textit{noAccountableSafety}$ does not have a model (unsatisfiable).
+Similar to Apalache, Alloy allows us to show safety for bounded scopes.
+
+\begin{lstlisting}[language=alloy,columns=fullflexible]
+  pred noAccountableSafety {
+    disagreement and (3.mul[#slashableNodes] < #Signature)
+  }
+  run noAccountableSafety for 10 but 6 Block, 6 Checkpoint,
+                                     4 Signature, 6 FfgVote, 15 Vote, 5 int
+\end{lstlisting}
+
+Our goal was to see, how far Alloy could go, when we were increasing the scope.
+Table~\ref{tab:alloy-mc} summarizes our experiments with Alloy when checking
+$\textit{noAccountableSafety}$ with Kissat. As we can see, we have managed to
+show accountable safety for the chain configurations of up to 6 blocks,
+including the genesis. However, the most interesting configuration of 7 blocks
+(on two chains) happens to be quite challenging for Alloy and Kissat.
```

Comment:
Why "(on two chains)"?
Doesn't this spec allow for arbitrary block graphs?

Reply:
True, this spec has an arbitrary graph